[core] Remove outdated Babel plugins

ZeeshanTamboli opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Part of #40958.

With the updated browser support, particularly for the new Safari browser, it seems we no longer require these plugins. Previous attempted PRs mentioned in #40958.

Additionally, it appears that the `assumptions` feature of Babel is also not needed. This closes #37461.

The plugins are now included by default in the latest version of `@babel/preset-env`. They are applied based on the browsers listed in `browserslist`, with Babel maintaining a mapping of versions for each plugin [here](https://github.com/babel/babel/blob/main/packages/babel-compat-data/data/plugins.json).

`browserstack-force` run for Safari: https://app.circleci.com/pipelines/github/mui/material-ui/128548/workflows/5eb395f3-c3f8-4e63-bdd3-f51a5e790419/jobs/693108

[Bundle size reduction](https://github.com/mui/material-ui/pull/42140#issuecomment-2095378382).

Janpot commented:
> packages/material-ui/material-ui.production.min.js: parsed: +18.78% , gzip: +5.84%

Any idea what's going on here?

ZeeshanTamboli commented:
> > packages/material-ui/material-ui.production.min.js: parsed: +18.78% , gzip: +5.84%
> 
> Any idea where this 18% increase comes from?

Not sure. It seems to be related to the `umd` build. Do you have any ideas on how we can compare the _actual_ build size of `master` versus the build on this branch for `umd`? It's possible that the `contributor-dashboard-legacy` code in `mui-public` is miscalculating it.

Janpot commented:
You could try running 
```
 pnpm --filter @mui/material build:umd
```
on master and copy the `packages/mui-material/build/umd/material-ui.production.min.js` somewhere else, then run the same command on this branch and then `git diff` both files . Maybe that will give a clue on what was added extra by this branch?

ZeeshanTamboli commented:
@Janpot Yesterday, I spent quite some time checking and debugging.

I compared the development files instead of the production ones because the production file is unreadable, likely due to terser. Here's what I did:
- Checked out the `next` branch.
- Ran `pnpm install` to get the Babel plugins.
- Built the UMD bundle using `pnpm --filter @mui/material build:umd`.
- Copied the `material-ui.development.js` file and renamed it to `next-material-ui.development.js`.
- Checked out this branch.
- Ran `pnpm install` and built the UMD bundle again.
- Copied the `material-ui.development.js` file and renamed it to `babel-branch-material-ui.development.js`.
- Compared both files using `git diff --no-index --word-diff next-material-ui.development.js babel-branch-material-ui.development.js`.

Here's a sample diff which is the same till the end: 

https://github.com/mui/material-ui/assets/20900032/01789cb9-cf46-4df3-82af-5f36e191f0bd



What I noticed is that it's using `objectSpread` instead of `_extends`, resulting in a larger bundle size. This change is due to the `@babel/plugin-transform-object-rest-spread` plugin removal. When `loose: true`, it uses the `_extends` helper (https://babeljs.io/docs/babel-plugin-transform-object-rest-spread#loose), which was the case in the `next` branch.

What I don't understand is why rollup's babel is using the `object-spread` plugin in the first place? Is it that it is not considering the `.browserslistrc` file and checking the browser versions?

Janpot reviewed (commented):
> Yesterday, I spent quite some time checking and debugging.

Uff, I'm sorry, that was not my intention. I believe the UMD bundle is up for [removal](https://github.com/mui/material-ui/issues/40960) in v6. We have visualized now the added code in the bundle, at least we now know it's in the bundling and not in the size calculation. Let's not waste more time on debugging this. For me this PR is good to merge after UMD is removed.

DiegoAndai commented:
@ZeeshanTamboli I plan to work on the UMD removal hopefully this week or next, but if you want to unblock this PR by removing it yourself, feel free to go ahead.

oliviertassinari commented:
How about we open a new issue for?

https://pagespeed.web.dev/analysis?url=https%3A%2F%2Fdeploy-preview-42140--material-ui.netlify.app%2F

<img width="933" alt="SCR-20240508-qkhd" src="https://github.com/mui/material-ui/assets/3165635/c89cffa0-be0a-482c-9531-bc39c99756a5">

This can be reproduced too in HEAD. I imagine that Lighthouse uses `Cannot call a class as a function` to detect https://babeljs.io/docs/babel-plugin-transform-classes. Yes, bingo: https://github.com/GoogleChrome/lighthouse/blob/369979f498bd6560127e10476edffb264d4fa3b9/core/audits/byte-efficiency/legacy-javascript.js#L301C22-L301C55. 

Seeing this on https://deploy-preview-42140--material-ui.netlify.app/ is wrong:

<img width="500" alt="SCR-20240508-tytg" src="https://github.com/mui/material-ui/assets/3165635/93f5b337-b64b-4a4f-b048-a0cf21575561">

We should at minimum go into loose mode: https://github.com/vercel/next.js/blob/5ff2731c589692ed86379f876a38e1ca46f5761e/packages/next/src/build/babel/preset.ts#L161.

ZeeshanTamboli commented:
> @ZeeshanTamboli I plan to work on the UMD removal hopefully this week or next, but if you want to unblock this PR by removing it yourself, feel free to go ahead.

@DiegoAndai Thanks. I've already begun working on this yesterday. Here's the PR: https://github.com/mui/material-ui/pull/42172. Feel free to push any changes if needed.

----

> How about we open a new issue for this?

@oliviertassinari Do we need to make any changes outside of this PR? I suppose it's all related to Next.js with the new issue you created there.

Janpot commented:
Given the Next.js track record with the maintenance of the babel setup in Next.js I would assume we'll be using their SWC setup before this gets fixed 😄.

Btw, yesterday, on top of this [PR](https://github.com/mui/mui-x/pull/13051) on the X repo I did a few quick tests and I could just remove the babel config without breaking their build. Page compilation in dev mode went from ~8s to ~4.5s on my machine. I couldn't enable turbopack due to our usage of `esmExternals`.

DiegoAndai commented:
@ZeeshanTamboli would it be possible to come up with a detailed list of Safari 12 features that will stop working on v6? This is so users can adjust accordingly.

Janpot reviewed (approved):
Good to go 👍
